pkg/cmd/roachtest/test.go

```diff
@@ -1163,16 +1163,33 @@ func (r *registry) runAsync(
 
 		timeout := c.expiration.Add(-10 * time.Minute).Sub(timeutil.Now())
 		if timeout <= 0 {
-			t.spec.Skip = fmt.Sprintf("cluster expired (%s)", timeout)
+			t.spec.Skip = fmt.Sprintf("cluster is already expired at beginning of test (%s)", timeout)
 			return
 		}
 
 		if t.spec.Timeout > 0 && timeout > t.spec.Timeout {
 			timeout = t.spec.Timeout
 		}
 
-		done := make(chan struct{})
-		defer close(done) // closed only after we've grabbed the debug info below
+		var shouldDestroyOnDefer bool
+		defer func() {
+			if !shouldDestroyOnDefer {
+				// If we're not supposed to destroy, at least stop the nodes.
+				// We've grabbed all the debug info we desired in the defer
+				// below this one already, and now the name of the game is
+				// getting the actual test code to return, which it is likely
+				// to do if the cluster it's running against stops running (on
+				// top of test's context already having been cancelled).
+				if c.nodes > 0 { // unit tests
+					_ = c.StopE(ctx, c.All())
+				}
+				return
+			}
+			// NB: c.destroyed is nil for cloned clusters (i.e. in subtests).
+			if !debugEnabled && c.destroyed != nil {
+				c.Destroy(ctx)
```

Comment:
PS I'm unclear why there's this random `c.Destroy` call that only fires on a timeout. Perhaps someone can educate me. I thought destroying would be left to a higher power. Perhaps we can just stop in all cases and don't need to track the bool?